channeldb/graph_test.go

```diff
@@ -1909,6 +1924,134 @@ func TestFetchChanInfos(t *testing.T) {
+// both sides.
+func TestIncompleteChannelPolicies(t *testing.T) {
+	t.Parallel()
+
+	db, cleanUp, err := makeTestDB()
+	defer cleanUp()
+	if err != nil {
+		t.Fatalf("unable to make test database: %v", err)
+	}
+
+	graph := db.ChannelGraph()
+
+	// Create two nodes.
+	node1, err := createTestVertex(db)
+	if err != nil {
+		t.Fatalf("unable to create test node: %v", err)
+	}
+	if err := graph.AddLightningNode(node1); err != nil {
+		t.Fatalf("unable to add node: %v", err)
+	}
+	node2, err := createTestVertex(db)
+	if err != nil {
+		t.Fatalf("unable to create test node: %v", err)
+	}
+	if err := graph.AddLightningNode(node2); err != nil {
+		t.Fatalf("unable to add node: %v", err)
+	}
+
+	// Create channel between nodes.
+	startTime := time.Unix(1234, 0)
+	endTime := startTime
+	txHash := sha256.Sum256([]byte{0})
+	op := wire.OutPoint{
+		Hash:  txHash,
+		Index: 0,
+	}
+
+	channel, chanID := createEdge(
+		uint32(0), 0, 0, 0, node1, node2,
+	)
+
+	if err := graph.AddChannelEdge(&channel); err != nil {
+		t.Fatalf("unable to create channel edge: %v", err)
+	}
+
+	// Ensure that channel is reported with unknown policies.
+	calls := 0
+	node2.ForEachChannel(nil, func(_ *bolt.Tx, _ *ChannelEdgeInfo,
+		outEdge, inEdge *ChannelEdgePolicy) error {
+		
+		if outEdge != nil {
+			t.Fatalf("Expected no outgoing policy for node2")
+		}
+
+		if inEdge != nil {
+			t.Fatalf("Expected no incoming policy for node2")
+		}
+
+		calls++
+
+		return nil
+	})
+
+	if calls != 1 {
+		t.Fatalf("Expected only one callback call for node2")
+	}
+
+	// Only create an edge policy for node1 and leave the policy for node2
+	// unknown.
+	updateTime := endTime
+	endTime = updateTime.Add(time.Second * 10)
+	edgePolicy := newEdgePolicy(
+		chanID.ToUint64(), op, db, updateTime.Unix(),
+	)
+	edgePolicy.Flags = 0
+	edgePolicy.Node = node2
+	edgePolicy.SigBytes = testSig.Serialize()
+	if err := graph.UpdateEdgePolicy(edgePolicy); err != nil {
+		t.Fatalf("unable to update edge: %v", err)
+	}
+
+	// Ensure that channel is reported as a channel of node2
+	calls = 0
+	node2.ForEachChannel(nil, func(_ *bolt.Tx, _ *ChannelEdgeInfo,
+		outEdge, inEdge *ChannelEdgePolicy) error {
+		
+		if outEdge != nil {
+			t.Fatalf("Expected no outgoing policy for node2")
+		}
+
+		if inEdge == nil {
+			t.Fatalf("Expected an incoming policy for node2")
+		}
+
+		calls++
+
+		return nil
+	})
+
+	if calls != 1 {
+		t.Fatalf("Expected only one callback call for node2")
+	}
+
+	// Ensure that channel is reported as a channel of node1
+	calls = 0
+	node1.ForEachChannel(nil, func(_ *bolt.Tx, _ *ChannelEdgeInfo,
+		outEdge, inEdge *ChannelEdgePolicy) error {
+		
+		if outEdge == nil {
+			t.Fatalf("Expected no outgoing policy for node1")
+		}
+
+		if inEdge != nil {
+			t.Fatalf("Expected an incoming policy for node1")
+		}
+
+		calls++
+
+		return nil
+	})
+
+	if calls != 1 {
+		t.Fatalf("Expected only one callback call for node1")
+	}
+}
```

Comment:
Excellent test! Could do a final check to ensure both policies are non-nil after adding the second one 😀

Reply:
Final check added.